add `MP_SMALL_STACK_SIZE` option

sjaeckel opened the pull request:
This adds an option to use a heap-buffer for the usually stack-based
`MP_WARRAY`-sized temporary buffers.

Per default it will reserve a single buffer, which can be modified
* at compile-time via the `MP_WARRAY_NUM` define
* at run-time by calling `mp_warray_init()`

The internal structure can only be created once. If one wants to modify
the maximum number of elements, the entire structure has to be free'd
by calling `mp_warray_free()`.

In case one wants to use this option with multiple threads, one shall
use the `mp_warray_init()` function and pass appropriate locking functions.

### Timings

| op / algo | ECC | RSA |
| ------------ | ------ | ------ |
| decrypt_key | ![ecc_decrypt_key](https://user-images.githubusercontent.com/294374/198363948-6ce27d0b-d60e-4db3-b173-a4c3cb7da8eb.svg) | ![rsa_decrypt_key](https://user-images.githubusercontent.com/294374/198363956-fcbb74cd-69ea-4206-bd44-f82db68eda3e.svg) |
| encrypt_key | ![ecc_encrypt_key](https://user-images.githubusercontent.com/294374/198363951-9aa0d363-436e-43fa-a08c-533809a53c43.svg) | ![rsa_encrypt_key](https://user-images.githubusercontent.com/294374/198363958-cfdfec24-e3f3-47a1-973e-f92b59f1b91b.svg) |
| sign_hash | ![ecc_sign_hash](https://user-images.githubusercontent.com/294374/198363954-dc1baf4e-7d05-4e0a-a165-4816bdf94de7.svg) | ![rsa_sign_hash](https://user-images.githubusercontent.com/294374/198363960-98389e64-635e-44d8-9e6b-4e6966d65b37.svg) |
| verify_hash | ![ecc_verify_hash](https://user-images.githubusercontent.com/294374/198363955-c3024a12-eece-4269-ac8d-1f9b9641ed6e.svg) | ![rsa_verify_hash](https://user-images.githubusercontent.com/294374/198363961-01a44c47-a00a-4ea8-a183-ea882016f308.svg) |


Related to: #441 #447 #511

sjaeckel committed:
add `MP_SMALL_STACK_SIZE` option

This adds an option to use a heap-buffer for the usually stack-based
`MP_WARRAY`-sized temporary buffers.

Per default it will reserve a single buffer, which can be modified
* at compile-time via the `MP_WARRAY_NUM` define
* at run-time by calling `mp_warray_init()`

The internal structure can only be created once. If one wants to modify
the maximum number of elements, the entire structure has to be free'd
by calling `mp_warray_free()`.

In case one wants to use this option with multiple threads, one shall
use the `mp_warray_init()` function and pass appropriate locking functions.

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel commented:
@jlaitine could you maybe have a look and tell me whether that's an acceptable solution for you?

czurnieden commented on doc/bn.tex:
```diff
@@ -428,7 +442,9 @@ \chapter{Getting Started with LibTomMath}
-libtommath.a).	There is no library initialization required and the entire library is thread safe.
+libtommath.a).	There is no library initialization required and the entire library is thread safe
+if it is used in its default configuration. Locking is recommended if the small-stack option
+is enabeled and multiple threads are used, c.f. \ref{ch:SMALL_STACK_INTRO} resp. \ref{ch:SMALL_STACK_API}
```
s/enabeled/enabled/

czurnieden commented on doc/bn.tex:
```diff
@@ -813,6 +829,53 @@ \subsection{Adding additional digits}
+\begin{alltt}
+mp_err mp_warray_init(size_t n_alloc, bool preallocate, mp_lock *lock);
+\end{alltt}
+
```
Maybe add a sentence about the flag `preallocate`? It is clear in the code but a bit ambiguous here.

czurnieden reviewed (changes requested):
Found nothing except some typos.

But the effect on the ECC computations *is* interesting!

czurnieden commented:
What happened to this PR? Did you change your mind? I find it quite interesting to have a choice between stack and heap for this. On the other side: it breaks our general thread-safety and needs locks.
But I'm still supportive.

sjaeckel commented on s_mp_warray_get.c:
```diff
@@ -0,0 +1,33 @@
+   }
+   if (s_mp_warray.allocated + 1 > s_mp_warray.usable)
+      goto LBL_OUT;
+   ret = MP_CALLOC(MP_WARRAY, sizeof(mp_word));
```
I guess we have to live with the worst-case assumption as the default value.

sjaeckel commented on .github/workflows/main.yml:
```diff
@@ -153,7 +159,7 @@ jobs:
         os: [ ubuntu-18.04, ubuntu-20.04 ]
         build_type: [ '', -DCMAKE_BUILD_TYPE=Debug, -DCMAKE_BUILD_TYPE=Release, -DCMAKE_BUILD_TYPE=RelWithDebInfo, -DCMAKE_BUILD_TYPE=MinSizeRel ]
-        cc: [ clang, gcc ]
+        cc: [ clang-10, gcc ]
```
I can't really remember, but maybe it had to do with the CI runners at the time of the original PR still being 18.04 and 20.04, where 18.04 default clang was too old or something like that.

Fixed.

sjaeckel commented:
> What happened to this PR? Did you change your mind?

No, I just didn't have the passion to follow up :)

> On the other side: it breaks our general thread-safety and needs locks.

That's indeed a thing. I'm not sure whether it'd make sense to replace the locks by taking care of that ourselves and using lock-free/atomic operations under the hood!?

sjaeckel committed:
Replace locking by atomic operations

`s_warray_init()` and `s_warray_free()` are not safe and MUST NOT be called
from multiple threads.

This also removes `MP_WARRAY_NUM`, since automatic initialization will not
be safe for more than one thread.

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
Add `s_mp_cmpexch_n()`

To be able to support this for MSVC as well, we have to move this into a
separate private API function.

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
Add multi-threaded tests

Output gets garbeled a bit, but we only care for the result which is
`Tests OK/NOP/FAIL: 50/0/0`.

Add `-Wno-incomplete-setjmp-declaration` since `clang-10` shipping with
Ubuntu 20.04 seems broken... and `-Wno-unknown-warning-option` since
`clang-8` doesn't know about this warning...

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel commented:
> I'm not sure whether it'd make sense to replace the locks by taking care of that ourselves and using lock-free/atomic operations under the hood!?

What do you think? Better like that!?

czurnieden commented:
Sorry, took a bit longer to check.

> Better like that!?

Yes, looks reasonable.

Caveat: I cannot test the Windows stuff now: my last PC here with Windows (what is the plural of Windows?) let the magic smoke out, need a visit to Ebay before going on.

PS: I am aware of my TODO list, it's not forgotten!

sjaeckel commented:
> what is the plural of Windows?

My SO just explained to me that the plural only works with the complete name! Microsoft Windows -> Microsoft**s** Windows

> my TODO list

No need to hurry, we have time.

> test the Windows stuff

I'll see what I can do by extending CI.

czurnieden commented:
> My SO just explained to me that the plural only works with the complete name! Microsoft Windows -> Microsofts Windows

Ah, ok, my thanks to your SO!

> I'll see what I can do by extending CI.

That would be really nice!

> No need to hurry, we have time.

You have, but I'm an overweight old fart already! ;-)

My main problem is the error I got with the base2->base10 conversion with a single round of NR to compute the next divisor at a very large (close to `MP_MAX_DIGIT_COUNT` with `MP_64BIT` ) number. Sadly a randomly generated large number and it hasn't been logged what it was (Full HD. Yes, I'm an idiot). Haven't been able to repeat the error and am trying to compute the accumulated error of the NR rounds now manually. Fiddly.

We need to be able to do at most 22 rounds with `n_0 = 1000` and the individual absolute error is about `2^(-34)`. But it is not floating point, it is integer math, so what is the accumulated error at the end? Does it even accumulate?

Was the aformentioned single error just a random bitflip? An error elsewhere?

\*hng\*!

sjaeckel committed:
Add tests for MSVC multi-threading

... and fix some MSVC related (and other) things.

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel commented:
> Ah, ok, my thanks to your SO!

You're very welcome ;-)

> > I'll see what I can do by extending CI.

wow, what a ride ... In between I was not sure whether I'd simply disable the feature for MSVC ... but now it seems to work^TM ...

sjaeckel committed:
Update makefiles

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel commented:
@jlaitine can you please check if this works for you?

sjaeckel committed:
Use Thread Local Storage for `warray` buffer

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
regen

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
Use appveyor build matrix

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
Run small-stack tests with valgrind

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel committed:
Disable MP_SMALL_STACK_SIZE on MSVC

Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>

sjaeckel commented:
> In between I was not sure whether I'd simply disable the feature for MSVC ... but now it seems to work^TM ...

aaaand in the end I disabled it for MSVC since IMO the most elegant solution is via TLS and MSVC seems to have a problem with TLS in DLLs. I'm not 100% sure what's going wrong, but since it was going fine with the static library [0] I suspect that we'd have to implement what's described in [1] and I'm simply not willing to do that, nor would I accept a PR that implements it.

If I'm mistaken there and someone with more MSVC experience wants to chime in and provide an easy fix, I'll happily accept that :)

[0] https://ci.appveyor.com/project/libtom/libtommath/builds/49507598
[1] https://learn.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library

czurnieden reviewed (approved):
Bearhug?

czurnieden commented:
> https://learn.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library

So that's the superlativ of "unnecessary complicated", isn't it?

jlaitine commented:
> @jlaitine can you please check if this works for you?

Hi sorry for the very long response time, I have been busy with some other things. I see that this already got merged, so I tested the development branch directly and this change works perfectly!

I still have a small compilation error, in PX4, which is compiled with "-Wall". Not related to this PR. I created anoter PR out of that: https://github.com/libtom/libtommath/pull/578

Thanks again,
Jukka

sjaeckel commented:
Thanks a lot for the feedback! Glad to hear that it works for you :)
